mantidproject/mantid
# Use H5Cpp library for LoadMcStas algorithm

### Description of work
This PR does some refactoring of the `LoadMcStas` algorithm so that it uses the HDF5 api directly to load the McStas simulation data rather than using the Nexus API. This results in a speed improvement of approximately 20%. This is the profile for loading one file before making the changes (925 seconds approx):
![image](https://github.com/user-attachments/assets/e8070c26-57ba-4a34-9c66-e14d851dac68)


This is the profile after the changes (725 seconds approx):
![image](https://github.com/user-attachments/assets/de67fa80-e3b4-41c5-9bbf-92b67a4c8430)


I believe a further speed improvement can be achieved by multi-threading the loading of each chunk of event data inside the `readEventData` function. I believe HDF5 makes this multi-threaded reading possible.

Part of #38356
Part of #37164

### To test:
1. Ask me to send you the data file (its large 12.9 GB)
1. Run the following script to load the *.h5 file (be warned it will take 700 seconds at least)
```python
from mantid.simpleapi import Load

Load(Filename='/home/mlc47243/Documents/other-documents/slow-workflows/mcstas/LR_2x2mm_SX_5e10_beam_stop_1/mccode.h5', OutputWorkspace='run_1')
```
3. There should be a group workspace containing the following workspaces:
![image](https://github.com/user-attachments/assets/6c8e4605-ddb0-4483-9200-cbb5c34728c4)
4. Plot the bin in the first workspace to get a plot that looks like this:
![image](https://github.com/user-attachments/assets/aa5b8d7b-117a-4061-9c6e-4174e830eadb)


### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
